src/Storages/ExternalStream/Kafka/KafkaSink.cpp

```diff
@@ -155,13 +159,22 @@ KafkaSink::KafkaSink(
     {
         /// The callback allows `IRowOutputFormat` based formats produce one Kafka message per row.
         writer = FormatFactory::instance().getOutputFormat(
-            data_format, *wb, header, context, [this](auto & /*column*/, auto /*row*/) { wb->next(); }, kafka.getFormatSettings(context));
+            data_format, *wb, header, context, [this](auto & /*column*/, auto /*row*/) { wb->markOffset(); wb->next(); }, kafka.getFormatSettings(context));
         if (!dynamic_cast<IRowOutputFormat*>(writer.get()))
-            throw Exception(ErrorCodes::INVALID_SETTING_VALUE, "Data format `{}` is not a row-based foramt, it cannot be used with `one_message_per_row`", data_format);
+            LOG_WARNING(logger, "Data format `{}` is not a row-based format, `one_message_per_row` setting will not be applied", data_format);
     }
     else
     {
-        writer = FormatFactory::instance().getOutputFormat(data_format, *wb, header, context, {}, kafka.getFormatSettings(context));
+        auto max_rows_per_message = context->getSettingsRef().kafka_max_message_rows.value;
+        writer = FormatFactory::instance().getOutputFormat(
+            data_format, *wb, header, context, [this, max_rows_per_message](auto & /*column*/, auto /*row*/) {
+                wb->markOffset();
+                if (++rows_in_current_message % max_rows_per_message == 0)
```

Comment:
nit : 
```
if (rows_in_current_message >= max_rows_per_message)
{
    rows_in_current_message = 0;
    ...
}
```

Reply:
Why only `++rows_in_current_message % max_rows_per_message == 0`, we do `wb->next()` ?

Reply:
This has already been handled in `addMessageToBatch` method. This counter needs to be reset when max size has reached as well.

Reply:
OutputFormat writer 
  -> writes to wb WriteBufferFromKafkaSink 
  -> for every write call back to WriteCallback which is the lambda function passed to `getOutputFormat`,
      the lamba function checks if we reach the `max_row_per_message` limit, if it does, call `wb->next()` which 
      in turn calls WriteBufferFromKafkaSink::nextImpl to addMessageToBatch(...), 
      look like the logic is broken here since we will add MessageToBatch until we reach the max limit ?

Reply:
> This has already been handled in `addMessageToBatch` method. This counter needs to be reset when max size has reached as well.

Yes, then let's use `>=` which is a bit more economic than '%`

Reply:
an ascii flow could be very helpful here btw. but in next pr

Reply:
`addMessageToBatch` is where it move the data out from the buffer (i.e. `WriteBufferFromKafkaSink`) to and create the Kafka message.

The output format serializes the rows in the chunk and write it to the buffer. The output format accepts a callback which will be called everything a row is written into the buffer. We use this callback to mark the offset and check if it reaches the row limit. If it reaches the row limit, it flush the buffer (by calling `wb->next()`).

Except reaching the row limit, there are two more cases that the buffer will flush:
1. All rows in the chunk have been written to the buffer.
2. The buffer is full.

So `addMessageToBatch` is the central place to handle all these flush logic. It creates the Kafka messages, and handle in-complete data (pending data), and reset the batch states. I believe there is nothing broken here.

Definitely worth improving the documentation to describe this logic better.